File: app/src/main/kotlin/net/tigr/soulcalc/domain/evaluator/Evaluator.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Albert Zenkoff
 */

package net.tigr.soulcalc.domain.evaluator

import net.tigr.soulcalc.domain.model.Result
import net.tigr.soulcalc.domain.model.Scope
import net.tigr.soulcalc.domain.parser.*
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * Evaluates AST nodes to produce numeric results.
 *
 * The evaluator walks the AST recursively, resolving variables and line
 * references from the provided scope, and computing the final numeric value.
 *
 * Note on scope mutation: The evaluator maintains a mutable scope internally
 * which is updated during evaluation of assignment expressions (e.g., `$x = 5`).
 * The updated scope is returned via [EvalResult.newScope], allowing callers to
 * propagate scope changes to subsequent evaluations. Each Evaluator instance
 * should be used for a single evaluation to avoid unexpected scope state.
 */
class Evaluator(private var scope: Scope = Scope()) {

    /**
     * Result of evaluation including potential scope changes from assignments.
     */
    data class EvalResult(
        val result: Result,
        val newScope: Scope
    )

    /**
     * Evaluates an AST node and returns the result along with any scope updates.
     */
    fun evaluate(node: AstNode): EvalResult {
        return try {
            val value = eval(node)
            EvalResult(Result.Success(value), scope)
        } catch (e: EvalException) {
            EvalResult(Result.Error(e.message ?: "Evaluation error"), scope)
        }
    }

    /**
     * Evaluates a parse result directly.
     */
    fun evaluate(parseResult: Parser.ParseResult): EvalResult {
        return when (parseResult) {
            is Parser.ParseResult.Success -> evaluate(parseResult.node)
            is Parser.ParseResult.Error -> EvalResult(Result.Error(parseResult.message), scope)
            is Parser.ParseResult.Empty -> EvalResult(Result.Empty, scope)
        }
    }

    /**
     * Returns the current scope (useful after evaluating assignments).
     */
    fun getScope(): Scope = scope

    private fun eval(node: AstNode): Double {
        return when (node) {
            is NumberNode -> node.value

            is BinaryOpNode -> evalBinaryOp(node)

            is UnaryMinusNode -> -eval(node.operand)

            is PercentNode -> evalPercent(node)

            is VariableNode -> evalVariable(node)

            is LineRefNode -> evalLineRef(node)

            is AssignmentNode -> evalAssignment(node)

            is FunctionNode -> evalFunction(node)
        }
    }

    private fun evalBinaryOp(node: BinaryOpNode): Double {
        val left = eval(node.left)
        val right = eval(node.right)

        return when (node.operator) {
            BinaryOp.ADD -> left + right
            BinaryOp.SUBTRACT -> left - right
            BinaryOp.MULTIPLY -> left * right
            BinaryOp.DIVIDE -> {
                if (right == 0.0) {
                    if (left == 0.0) {
                        throw EvalException("NaN")
                    }
                    throw EvalException(if (left < 0) "-∞" else "∞")
                }
                left / right
            }
            BinaryOp.POWER -> left.pow(right)
        }
    }

    private fun evalPercent(node: PercentNode): Double {
        val operandValue = eval(node.operand)
        val percentage = operandValue / 100.0

        return if (node.base != null) {
            val baseValue = eval(node.base)
            baseValue * percentage
        } else {
            percentage
        }
    }

    private fun evalVariable(node: VariableNode): Double {
        return scope.resolveVariable(node.name)
            ?: throw EvalException("? \$${node.name}")
    }

    private fun evalLineRef(node: LineRefNode): Double {
        return scope.resolveLineRef(node.lineNumber)
            ?: throw EvalException("? \$${node.lineNumber}")
    }

    /**
     * Evaluates an assignment and updates the internal scope.
     * The updated scope is accessible via [EvalResult.newScope] after evaluation completes.
     */
    private fun evalAssignment(node: AssignmentNode): Double {
        val value = eval(node.expression)
        scope = scope.withVariable(node.variableName, value)
        return value
    }

    private fun evalFunction(node: FunctionNode): Double {
        val argValue = eval(node.argument)

        return when (node.name.lowercase()) {
            "sqrt" -> {
                if (argValue < 0) {
                    throw EvalException("NaN")
                }
                sqrt(argValue)
            }
            else -> throw EvalException("Unknown function: ${node.name}")
        }
    }

    private class EvalException(message: String) : Exception(message)
}

/**
 * Convenience function to evaluate an expression string.
 */
fun evaluateExpression(input: String, scope: Scope = Scope()): Evaluator.EvalResult {
    val parseResult = parseExpression(input)
    return Evaluator(scope).evaluate(parseResult)
}

File: ios/SoulCalc/Domain/Evaluator/Evaluator.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Albert Zenkoff

import Foundation

/// Result of evaluation including potential scope changes from assignments.
struct EvalResult {
    let result: Result
    let newScope: Scope
}

/// Evaluates AST nodes to produce numeric results.
///
/// The evaluator walks the AST recursively, resolving variables and line
/// references from the provided scope, and computing the final numeric value.
class Evaluator {
    private var scope: Scope

    init(_ scope: Scope = Scope()) {
        self.scope = scope
    }

    /// Evaluates an AST node and returns the result along with any scope updates.
    func evaluate(_ node: AstNode) -> EvalResult {
        do {
            let value = try eval(node)
            return EvalResult(result: .success(value), newScope: scope)
        } catch let error as EvalError {
            return EvalResult(result: .error(error.message), newScope: scope)
        } catch {
            return EvalResult(result: .error("Evaluation error"), newScope: scope)
        }
    }

    /// Evaluates a parse result directly.
    func evaluate(_ parseResult: ParseResult) -> EvalResult {
        switch parseResult {
        case .success(let node):
            return evaluate(node)
        case .error(let message, _):
            return EvalResult(result: .error(message), newScope: scope)
        case .empty:
            return EvalResult(result: .empty, newScope: scope)
        }
    }

    /// Returns the current scope.
    func getScope() -> Scope { scope }

    private func eval(_ node: AstNode) throws -> Double {
        switch node {
        case .number(let value):
            return value

        case .binaryOp(let left, let op, let right):
            return try evalBinaryOp(left: left, op: op, right: right)

        case .unaryMinus(let operand):
            return try -eval(operand)

        case .percent(let operand, let base):
            return try evalPercent(operand: operand, base: base)

        case .variable(let name):
            return try evalVariable(name)

        case .lineRef(let lineNumber):
            return try evalLineRef(lineNumber)

        case .assignment(let variableName, let expression):
            return try evalAssignment(variableName: variableName, expression: expression)

        case .function(let name, let argument):
            return try evalFunction(name: name, argument: argument)
        }
    }

    private func evalBinaryOp(left: AstNode, op: BinaryOp, right: AstNode) throws -> Double {
        let leftVal = try eval(left)
        let rightVal = try eval(right)

        switch op {
        case .add:
            return leftVal + rightVal
        case .subtract:
            return leftVal - rightVal
        case .multiply:
            return leftVal * rightVal
        case .divide:
            if rightVal == 0.0 {
                if leftVal == 0.0 {
                    throw EvalError("NaN")
                }
                throw EvalError(leftVal < 0 ? "-∞" : "∞")
            }
            return leftVal / rightVal
        case .power:
            return pow(leftVal, rightVal)
        }
    }

    private func evalPercent(operand: AstNode, base: AstNode?) throws -> Double {
        let operandValue = try eval(operand)
        let percentage = operandValue / 100.0

        if let base = base {
            let baseValue = try eval(base)
            return baseValue * percentage
        } else {
            return percentage
        }
    }

    private func evalVariable(_ name: String) throws -> Double {
        guard let value = scope.resolveVariable(name) else {
            throw EvalError("? $\(name)")
        }
        return value
    }

    private func evalLineRef(_ lineNumber: Int) throws -> Double {
        guard let value = scope.resolveLineRef(lineNumber) else {
            throw EvalError("? $\(lineNumber)")
        }
        return value
    }

    private func evalAssignment(variableName: String, expression: AstNode) throws -> Double {
        let value = try eval(expression)
        scope = scope.withVariable(variableName, value)
        return value
    }

    private func evalFunction(name: String, argument: AstNode) throws -> Double {
        let argValue = try eval(argument)

        switch name.lowercased() {
        case "sqrt":
            if argValue < 0 {
                throw EvalError("NaN")
            }
            return Foundation.sqrt(argValue)
        default:
            throw EvalError("Unknown function: \(name)")
        }
    }
}

/// Error thrown during evaluation.
private struct EvalError: Error {
    let message: String

    init(_ message: String) {
        self.message = message
    }
}

/// Convenience function to evaluate an expression string.
func evaluateExpression(_ input: String, scope: Scope = Scope()) -> EvalResult {
    let parseResult = parseExpression(input)
    return Evaluator(scope).evaluate(parseResult)
}
